gym/envs/registration.py

```diff
@@ -680,63 +310,168 @@ def make(id: Literal[
     "Ant-v2"
 ], **kwargs) -> Env[np.ndarray, np.ndarray]: ...
 
-# ----------------------------------------
 
 @overload
 def make(id: str, **kwargs) -> Env: ...
+@overload
+def make(id: EnvSpec, **kwargs) -> Env: ...
+
 # fmt: on
-def make(id: str, **kwargs) -> Env:
-    return registry.make(id, **kwargs)
 
 
-def spec(id: str) -> EnvSpec:
-    return registry.spec(id)
+# Global registry of environments. Meant to be accessed through `register` and `make`
+registry: dict[str, EnvSpec] = dict()
+current_namespace: Optional[str] = None
+
+
+def check_spec_register(spec: EnvSpec):
+    global registry, current_namespace
+    if current_namespace is not None:
+        if spec.namespace is not None:
+            logger.warn(
+                f"Custom namespace `{spec.namespace}` is being overridden "
+                f"by namespace `{current_namespace}`. If you are developing a "
+                "plugin you shouldn't specify a namespace in `register` "
+                "calls. The namespace is specified through the "
+                "entry point package metadata."
+            )
+
+    latest_versioned_spec = max(
+        (
+            spec_
+            for spec_ in registry.values()
+            if spec_.name == spec.name and spec_.version is not None
+        ),
+        key=lambda spec_: int(spec_.version),  # type: ignore
+        default=None,
+    )
+
+    unversioned_spec = next(
+        (
+            spec_
+            for spec_ in registry.values()
+            if spec_.namespace == spec.namespace
+            and spec_.name == spec.name
+            and spec_.version is None
+        ),
+        None,
+    )
+
+    if unversioned_spec and spec.version is not None:
+        raise error.RegistrationError(
+            "Can't register the versioned environment "
+            f"`{spec.id}` when the unversioned environment "
+            f"`{unversioned_spec.id}` of the same name already exists."
+        )
+    elif latest_versioned_spec and spec.version is None:
+        raise error.RegistrationError(
+            "Can't register the unversioned environment "
+            f"`{spec.id}` when the versioned environment "
+            f"`{latest_versioned_spec.id}` of the same name "
+            f"already exists. Note: the default behavior is "
+            f"that `gym.make` with the unversioned environment "
+            f"will return the latest versioned environment"
+        )
+
+
+# Public API
 
 
 @contextlib.contextmanager
 def namespace(ns: str):
-    with registry.namespace(ns):
-        yield
+    global current_namespace
+    old_namespace = current_namespace
+    current_namespace = ns
+    yield
+    current_namespace = old_namespace
+
+
+def register(id: str, **kwargs):
```

Comment:
public function w/o docstring

Reply:
Good catch, I'll add that (and the others)